autowire/shared/src/main/scala/autowire/Macros.scala

```diff
@@ -15,42 +14,44 @@ object Macros {
     def withFilter(f: T => Boolean): Check[T]
   }
   case class Luz[T](s: String) extends Check[T] {
-    def map[V](f: T => V) = Luz[V](s)
-    def flatMap[V](f: T => Check[V]) = Luz[V](s)
-    def withFilter(f: T => Boolean) = Luz[T](s)
+    def map[V](f: T => V): Luz[V] = Luz[V](s)
+    def flatMap[V](f: T => Check[V]): Luz[V] = Luz[V](s)
+    def withFilter(f: T => Boolean): Luz[T] = Luz[T](s)
   }
 
   case class Win[T](t: T, s: String) extends Check[T] {
-    def map[V](f: T => V) = Win(f(t), s)
-    def flatMap[V](f: T => Check[V]) = f(t)
-    def withFilter(f: T => Boolean) = if (f(t)) this else Luz(s)
+    def map[V](f: T => V): Win[V] = Win(f(t), s)
+    def flatMap[V](f: T => Check[V]): Check[V] = f(t)
+    def withFilter(f: T => Boolean): Check[T] with Product = if (f(t)) this else Luz(s)
   }
 
 
-  class MacroHelp[C <: Context](val c: C) {
+  class MacroHelp[C <: blackbox.Context](val c: C) {
+
     import c.universe._
-    def futurize(t: Tree, member: MethodSymbol) = {
+
+    def futurize(t: Tree, member: MethodSymbol): Tree = {
       if (member.returnType <:< c.typeOf[Future[_]]) t
       else q"scala.concurrent.Future($t)"
     }
 
-    def getValsOrMeths(curCls: Type): Iterable[Either[(c.Symbol, MethodSymbol), (c.Symbol, MethodSymbol)]] = {
+    def getValsOrMeths(curCls: Type): Iterable[Either[(Symbol, MethodSymbol), (Symbol, MethodSymbol)]] = {
       def isAMemberOfAnyRef(member: Symbol) = weakTypeOf[AnyRef].members.exists(_.name == member.name)
-      val membersOfBaseAndParents: Iterable[Symbol] = curCls.declarations ++ curCls.baseClasses.map(_.asClass.toType.declarations).flatten
-      val extractableMembers = for {
+      val membersOfBaseAndParents: Iterable[Symbol] = curCls.decls ++ curCls.baseClasses.flatMap(_.asClass.toType.decls)
+
+      val extractableMembers: Iterable[(Symbol, MethodSymbol)] = for {
         member <- membersOfBaseAndParents
         if !isAMemberOfAnyRef(member)
         if !member.isSynthetic
         if member.isPublic
         if member.isTerm
-        if member.isAbstract
         memTerm = member.asTerm
         if memTerm.isMethod
       } yield {
         member -> memTerm.asMethod
       }
 
-      extractableMembers flatMap { case (member, memTerm) =>
+      extractableMembers.toList.distinct flatMap { case (member, memTerm) =>
```

Comment:
No need to materialise code from duplicate method names (one from api, one from implementation). This will cut materialised code almost in half.